keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt

```diff
@@ -0,0 +1,38 @@
+package com.netflix.spinnaker.keel.persistence
+
+import com.netflix.spinnaker.keel.notifications.Notifier
+import org.springframework.beans.factory.annotation.Value
+import java.time.Clock
+
+/**
+ * A repository for storing a list of ongoing notifications, and calculating whether
+ * they should be resent
+ */
+abstract class NotifierRepository(
+  open val clock: Clock,
+  @Value("notify.waiting-duration") var waitingDuration: String = "P1D"
+) {
+
+  /**
+   * Adds a notification to the list of ongoing notifications.
+   * Assumption: each notifier sends only one type of message
+   * @return true if we should notify right now
+   */
+  abstract fun addNotification(resourceId: String, notifier: Notifier): Boolean
+
+  /**
+   * Clears a notification from the list of ongoing notifications.
+   * Does nothing if notification does not exist.
+   */
+  abstract fun clearNotification(resourceId: String, notifier: Notifier)
```

Comment:
P.S. The way this would feel more natural to me would be something like this:

```kotlin
enum class NotificationType {
  UNHEALTHY_RESOURCE
}

enum class NotificationScope {
  RESOURCE,
  // ...
}

abstract class Notification(
  val type: NotificationType,
  val scope: NotificationScope,
  val entityId: String, // resource ID, environment ID, etc.
  val subject: String,
  val message: String
)

data class ResourceUnhealthyNotification(
  val resource: Resource<*>
) : Notification(NotificationType.RESOURCE_UNHEALTHY, NotifiationScope.RESOURCE, resource.id, "Unhealthy resource", "Resource ${resource.id} is unhealthy")
```

Then your repository methods would look like:
```kotlin
abstract fun addNotification(notification: Notification)
```

So, actually no need for generics, just a bit of inheritance.

Reply:
So we aren't saving the notification, we are just relying on the message to keep being sent via the event. This class is essentially throttling the message (it's saying, "we can't send another message until 1 day has passed, I'll throw this message out"). The other way to do things would be to save the notification in the database (like you suggested) and then iterate through the saved notifications and send them if it's time. I opted not to do this because there is no way to not send the event every time, so it seemed useless to save the message.

Basically, that means if you want to send two different types of "unhealthy" notifications for the same entity you'll need two different `Notifier`s, because we throttle based on entity id + notifier type (we do not take into account the message content at all).

Does that help?